network/src/protocols/wire/handshake/v1/mod.rs

```diff
@@ -31,6 +31,23 @@ pub enum ProtocolId {
     OnchainDiscoveryRpc = 7,
 }
 
+impl fmt::Display for ProtocolId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use ProtocolId::*;
+        let descriptor = match self {
+            ConsensusRpc => "ConsensusRpc",
+            ConsensusDirectSend => "ConsensusDirectSend",
+            MempoolDirectSend => "MempoolDirectSend",
+            StateSynchronizerDirectSend => "StateSynchronizerDirectSend",
+            DiscoveryDirectSend => "DiscoveryDirectSend",
+            HealthCheckerRpc => "HealthCheckerRpc",
+            IdentityDirectSend => "IdentityDirectSend",
+            OnchainDiscoveryRpc => "OnchainDiscoveryRpc",
+        };
+        write!(f, "{}", descriptor)
+    }
+}
```

Comment:
suggestion: I think we can do something like we did w/ `RoleType` (https://github.com/libra/libra/blob/master/config/src/config/mod.rs#L121):

```rust
impl ProtocolId {
    #[inline]
    fn as_str(&self) -> &'static str {
        use ProtocolId::*;
        match self {
            ConsensusRpc => "ConsensusRpc",
            // ..
        }
    }
}
```
and then impl Display using `as_str()`. however, instead of using `protocol.to_string()` when recording metrics, we can instead use `protocol.as_str()` which should avoid some unnecessary allocation (and should be inlined).

Reply:
maybe some day we can write a little macro to do this for us 😅

Reply:
Okay, I'll change that. Btw, I think that the impl for `RoleType` should take self by reference?

Reply:
huh that is kinda weird... it looks like it works b/c `RoleType` is `Copy`